FIX: Avoid making new created assets dirty when opened

jfreire-unity opened the pull request:
### Description

https://jira.unity3d.com/browse/ISX-1591

### Changes made

Assets without any action maps and control schemes are not set as "dirty" anymore.

### Notes

@Pauliusd01 since you reported this, please try it out and check if it solves the reported issue.

### Checklist

Before review:

- [ ] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - FogBugz ticket attached, example `([case %number%](https://issuetracker.unity3d.com/issues/...))`.
    - FogBugz is marked as "Resolved" with *next* release version correctly set.
- [ ] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.
- [ ] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

During merge:

- [ ] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

jfreire-unity committed:
Avoid new created assets to be dirty when opened

ritamerkl commented:
LGTM, only one thing: what happens if you have an existing asset and delete all action maps and control schemes? should still be dirty then, right? (it's a edge case, but still worth trying I think)

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/InputActionsEditorWindow.cs:
```diff
@@ -210,6 +210,11 @@ private void DirtyInputActionsEditorWindow(InputActionsEditorState newState)
+                return false;
+
             var newAssetJson = asset.ToJson();
             return newAssetJson != m_AssetJson;
```
I would recommend determining (if not really checked) why the this evaluation evaluates to true instead of false (?) instead of introducing additional conditions for some corner case that shouldn't exist if the underlying asset json produces different or similar result. Isn't the asset json for the newly created asset "empty" similar to what asset.ToJson() would produce or is there any difference there to look into and resolve?

jfreire-unity replied:
Currently, the asset on disk has a specific layout.
Ideally, the layout shouldn't be `"{}"`. It should be a JSON serialized version of an empty asset. 
However, the `name` field is the problem. If you create and rename an input asset, its `name` field inside the asset is not updated at that stage. When opening the asset, there are actually changes to be made (the asset is flagged dirty) since the filename was changed.
The reason why this seems to happen is because on importing an asset, [we're actually setting `asset.name` based on the filename.](https://github.com/Unity-Technologies/InputSystem/blob/b0bcfbfbd0aa0561069447e1ec76d2ba7ec042fa/Packages/com.unity.inputsystem/InputSystem/Editor/AssetImporter/InputActionImporter.cs#L85-L87)

I spent some time on this and am feeling it can become a rabbit hole and introduce regressions. But I think we should revisit this workflow.

ekcoh replied:
Sounds odd we are setting name on import and not directly after constructing the asset, e.g. create, import, set name, write if needed, but I think you know best what is the best way to resolve this.

Pauliusd01 commented:
> LGTM, only one thing: what happens if you have an existing asset and delete all action maps and control schemes? should still be dirty then, right? (it's a edge case, but still worth trying I think)

Good catch, this indeed does not trigger the dirty state even though it should. @jfreire-unity

Pauliusd01 reviewed (changes requested):
Updating status to waiting for fixes

jfreire-unity committed:
Add methods to check asset for default json layout and being empty

jfreire-unity committed:
Improve dirty asset detection when new asset is first created

Ideally at this stage, the file on disk should have a serializable equivalent as the serialized asset. However, because the new asset on disk is being created with a specific default json layout, the serialized JSON asset doesn't have not the same content.
There are opportunities for change in this area.

jfreire-unity commented:
@Pauliusd01 thanks! you can try again

jfreire-unity commented:
> LGTM, only one thing: what happens if you have an existing asset and delete all action maps and control schemes? should still be dirty then, right? (it's a edge case, but still worth trying I think)

Yep, quite on point use case. Thanks! I fixed it now

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/InputActionsEditorWindow.cs:
```diff
@@ -210,6 +210,14 @@ private void DirtyInputActionsEditorWindow(InputActionsEditorState newState)
+            // If it is, there's nothing to save.
+            // At the moment, an asset only has the default asset layout content on disk when it is first created.
+            // So in this case we cannot go through the normal path and compare what's on disk with what has been serialized.
+            if (InputActionAsset.HasDefaultJsonLayout(m_AssetJson) && asset.IsEmpty())
```
Isn't this specific to this editor? I would recommend moving the logic to the associated utils class or making it an editor-only extension for the asset?

jfreire-unity replied:
Thanks. Remove the method and just have it inline as I don't know if it needs to be used elsewhere at the moment.
Checking if the asset is empty is not editor dependent.

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Actions/InputActionAsset.cs:
```diff
@@ -83,6 +83,8 @@ public class InputActionAsset : ScriptableObject, IInputActionCollection2
         /// InputActionAssets.
         /// </remarks>
         public const string Extension = "inputactions";
+        ////REVIEW: actually pre-populate with some stuff?
```
What is the content after a save of an empty asset? Not sure why do not initialise it like that when its created? Then we wouldn't need these checks at all and the equality check would solve that for us

jfreire-unity replied:
Check https://github.com/Unity-Technologies/InputSystem/pull/1761#discussion_r1345711499

jfreire-unity committed:
Remove default json layout check method

jfreire-unity committed:
Update renamed constant

Pauliusd01 reviewed (approved):
LGTM, did not notice anymore unusual scenarios where the asset gets dirty when it shouldn't (when undoing changes, redoing, creating new ones, cancelling already made ones and seeing if they are still dirty, etc)

jfreire-unity committed:
Merge branch 'develop' into isx-1591-make-new-empty-assets-not-dirty

jfreire-unity committed:
Merge branch 'develop' into isx-1591-make-new-empty-assets-not-dirty

jfreire-unity committed:
Merge branch 'develop' into isx-1591-make-new-empty-assets-not-dirty
